### ISSUE TITLE
day duration leads to invalid local time

### ISSUE THREAD

**Issue by autarch** [houseabsolute/DateTime.pm#21]
Migrated from [rt.cpan.org #87769](https://rt.cpan.org/Ticket/Display.html?id=87769) (status was 'new')

Requestors:
* zefram@fysh.org

From zefram@fysh.org on 2013-08-11 16:02:54:

With DateTime-1.03:

```
$dt0 = DateTime->new(year => 2013, month => 3, day => 30, hour => 1, time_zone => "Europe/London");
$dt1 = DateTime->new(year => 2013, month => 3, day => 31, hour => 2, time_zone => "Europe/London");
$dur = $dt1 - $dt0;
# $dur is 1 day
print $dt0 + $dur;
# => Invalid local time for date in time zone: Europe/London
```

As with the previous bug, this should be an easy case for duration
arithmetic, because the duration is being used additively to represent
the same calendar interval from which it was generated.  $dur is
sane in describing the interval as 1 day; the two times are exactly 24
hours apart.  However, adding that day onto $dt0 requires accounting for
the DST transition at the target time.  ->add_duration is not in fact
doing what ->subtract_datetime assumed it would, and so the addition
and subtraction operations are mismatched.

-zefram

**Comment by danstewart** [comment]
The error is coming from https://github.com/houseabsolute/DateTime-TimeZone/blob/master/lib/DateTime/TimeZone.pm#L233

March 31st 2013 was when the clocks changed:
https://www.timeanddate.com/time/change/uk/london?year=2013

Will have a look at it further to see if anything can be done to fix it but I assume there's a reason it currently dies.

**Comment by autarch** [comment]
Yeah, the problem is that date math can lead to invalid local times. There's a whole long section in the DateTime docs that should cover all the cases and what the module does in each case.

**Comment by autarch** [comment]
Closing this since this is the documented behavior - https://metacpan.org/pod/DateTime#How-DateTime-Math-Works
